torch/_subclasses/fake_tensor.py

```diff
@@ -1414,6 +1524,215 @@ def __exit__(self, a, b, c):
+            func.name(), torch._C.DispatchKey.CompositeImplicitAutograd
+        ):
+            raise _BypassDispatchCache("CompositeImplicitAutograd")
+
+        key_values = (
+            func,
+            # Translate any FakeTensor args to metadata.
+            self._translate_arg(args),
+            self._translate_arg(kwargs),
+            # Capture the default_dtype mode since that can affect the output tensor,
+            # e.g., when operating on constant float values.
+            torch.get_default_dtype(),
+            # The disallowance of dynamic shapes can affect caching, e.g., a change of
+            # mode can introduce a DynamicOutputShapeException where it wasn't seen on
+            # a previous instance of the same func and arg combination.
+            self.shape_env and self.shape_env.allow_dynamic_output_shape_ops,
+        )
+        return _DispatchCacheKey(key_values)
+
+    def _translate_arg(self, arg: Any) -> Any:
+        """
+        Translate the provided arg into a form suitable for caching at FakeTensor
+        dispatch, i.e., convert unhashable types like lists & dicts into tuples and
+        convert FakeTensors into metadata. Raises _BypassDispatchCache to signal
+        unsupported cases that should bypass caching.
+        """
+        if isinstance(arg, (list, tuple)):
+            return tuple(self._translate_arg(x) for x in arg)
+        elif isinstance(arg, dict):
+            return tuple((k, self._translate_arg(v)) for k, v in arg.items())
+        elif isinstance(arg, (torch.SymBool, torch.SymInt, torch.SymFloat)):
+            raise _BypassDispatchCache("symbolic shape")
+        elif isinstance(arg, FakeTensor):
+            if not self.is_our_fake(arg):
+                raise _BypassDispatchCache("not our fake")
+            if arg._has_symbolic_sizes_strides:
+                raise _BypassDispatchCache("symbolic shape")
+            if arg.constant is not None:
+                raise _BypassDispatchCache("constant attribute")
+            return extract_tensor_metadata(arg)
+        elif isinstance(arg, torch.Tensor):
+            raise _BypassDispatchCache("non-fake tensor")
+        else:
+            # It's important to capture the type of the arg since, e.g., 1 and 1.0
+            # hash to the same value, but can produce different dtypes for the
+            # output tensor.
+            return (type(arg), arg)
+
+    def _create_cache_entry(
+        self, key: _DispatchCacheKey, args: Tuple[Any], output: FakeTensor
+    ) -> _DispatchCacheEntry:
+        """
+        Create an entry in the cache for the given 'output' Tensor.
+        """
+        # Some ops return tuples of Tensors, but it's rare, so avoid
+        # the complexity of caching other types.
+        if not isinstance(output, FakeTensor):
+            raise _BypassDispatchCache("non-FakeTensor output")
+
+        # Avoid caching FakeTensors with constants attached since those
+        # can be invalidated.
```

Comment:
@eellison I heard you might know something about the constant field that can be attached to the FakeTensor. Am I right that it would be challenging to support correctly with this FakeTensor cache?

Reply:
Hmm, it's at least more challenging, yea. How often our we getting cache misses because of it ? I think definitely in the first pr not needed. If it's a huge source of misses we can look into it.

Reply:
Well on Chillee's gpt-fast at least, it doesn't show up:

hits = 22,435
misses = 523
bypasses = {
  'dynamic output shape': 8,
  'CompositeImplicitAutograd': 306,
  'non-FakeTensor output': 107,
  'lift': 60,
  'non-fake tensor': 35,
}

Reply:
Yea, let's not handle it for now